plugins/rest/auth.go

```diff
@@ -532,17 +549,21 @@ func (acs *awsCredentialServiceChain) addService(service awsCredentialService) {
 	acs.awsCredentialServices = append(acs.awsCredentialServices, service)
 }
 
-func (acs *awsCredentialServiceChain) credentials() (aws.Credentials, error) {
+func (acs *awsCredentialServiceChain) credentials(ctx context.Context) (aws.Credentials, error) {
 	for _, service := range acs.awsCredentialServices {
-		credential, err := service.credentials()
-		if err == nil {
-			acs.logger.Debug("awsSigningAuthPlugin:%s successful",
-				reflect.TypeOf(service).String())
-			return credential, nil
+		credential, err := service.credentials(ctx)
+		if err != nil {
+			acs.logger.Debug("awsSigningAuthPlugin:%T failed: %v", service, err)
+
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return aws.Credentials{}, err
+			}
+
+			continue
```

Comment:
Can you please help me understand the behavior here? Now we don't return the first error.

Reply:
We didn't return the first error before either.
The `awsCredentialServiceChain` basically iterates through all services and returns credentials from the first service that succeeds. Errors are ignored (but logged) and the `credentials()` call only fails if none of the chained services succeed.